lib/execution_engine2/sdk/EE2Runjob.py

```diff
@@ -457,6 +498,189 @@ def _check_job_arguments(self, jobs, has_parent_job=False):
+        if job.retry_parent:
+            return self.retry(str(job.retry_parent), as_admin=as_admin)
+
+        # Get run job params from db, and inject parent job id, then run it
+        run_job_params = self._get_run_job_params_from_existing_job(
+            job, user_id=self.sdkmr.user_id
+        )
+        # Submit job to job scheduler or fail and not count it as a retry attempt
+        run_job_params[_PARENT_RETRY_JOB_ID] = job_id
+        retry_job_id = self.run(params=run_job_params, as_admin=as_admin)
+
+        # Save that the job has been retried, and increment the count. Notify the parent(s)
+        # 1) Notify the parent container that it has a new child..
+        if parent_job:
+            try:
+                parent_job.modify(push__child_jobs=retry_job_id)
+            except Exception as e:
+                self._db_update_failure(
+                    job_that_failed_operation=str(parent_job.id),
+                    job_to_abort=retry_job_id,
+                    exception=e,
+                )
+
+        # 2) Notify the retry_parent that it has been retried
+        try:
+            job.modify(inc__retry_count=1)
+        except Exception as e:
+            self._db_update_failure(
+                job_that_failed_operation=str(job.id),
+                job_to_abort=retry_job_id,
+                exception=e,
+            )
+
+        # Should we compare the original and child job to make sure certain fields match,
+        # to make sure the retried job is correctly submitted? Or save that for a unit test?
+        return {"job_id": job_id, "retry_id": retry_job_id}
+
+    def retry(self, job_id: str, as_admin=False) -> Dict[str, Optional[str]]:
+        """
+        #TODO Add new job requirements/cgroups as an optional param
+        :param job_id: The main job to retry
+        :param as_admin: Run with admin permission
+        :return: The child job id that has been retried
+        """
+        job, parent_job = self._validate_retry_presubmit(
+            job_id=job_id, as_admin=as_admin
+        )
+        return self._retry(
+            job_id=job_id, job=job, parent_job=parent_job, as_admin=as_admin
+        )
+
+    def retry_multiple(
+        self, job_ids, as_admin=False
+    ) -> List[Dict[str, Union[str, Any]]]:
+        """
+        #TODO Add new job requirements/cgroups as an optional param
+        #TODO Notify the parent container that it has multiple new children, instead of multiple transactions?
+
+        :param job_ids: The list of jobs to retry
+        :param as_admin: Run with admin permission
+        :return: The child job ids that have been retried or errors
+        """
+        if not job_ids:
+            raise ValueError("No job_ids provided to retry")
+
+        # Check all inputs before attempting to start submitting jobs
+        retried_jobs = []
+        jobs = []
+        parent_jobs = []
+        for job_id in job_ids:
```

Comment:
I would probably note here that `_validate` doesn't do a recursive check if if the job has a retry parent, but the `_validate` call on the recursion is guaranteed to pass because the parent was retried once already so the `_validate` must have passed previously. Since the parent job's state can't have changed it would just pass again.
Not super important but I spent a few minutes thinking this through so it'd be helpful for the next person that comes along.

Reply:
added

Reply:
:+1: BTW I'm assuming here that my logic is correct, if I'm wrong let me know